custom-routes-vis/src/main/java/com/google/sps/servlets/GetDirectionsServlet.java

```diff
@@ -0,0 +1,120 @@
+    String destination = request.getParameter("destination");
+    String endpoint = request.getParameter("endpoint");
+
+    String apiKey = request.getParameter("apiKey");
+    String directionsApiKey = System.getenv("DIRECTIONS_API_KEY");
+    String key = apiKey == "" ? apiKey : directionsApiKey;  // This works but it seems backwards to me.
+
+    URLConnection connection;
+
+    switch (endpoint) {
+      case "compute-routes":
+        connection = sendPostRequestToRoutesApi(origin, destination, false, key);
+        break;
+      case "compute-routes-alpha":
+        connection = sendPostRequestToRoutesApi(origin, destination, true, key);
+        break;
+      default:  // Default to use the Directions API.
+        connection = sendGetRequestToDirectionsApi(origin, destination, key);
+        break;
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(readResponse(connection));
+  }
+
+  /** 
+   * Sends request to the Directions API and returns the connection.
+   * @param origin Origin of the requested routes in "lat,lng" format.
+   * @param destination Destination of the requested routes in "lat,lng" format.
+   * @param key The API key.
+   */
+  public URLConnection sendGetRequestToDirectionsApi(String origin, String destination, String key) 
+      throws IOException, MalformedURLException {
+    String originDestination = "&origin=" + origin + "&destination=" + destination;
+    URL directionsUrl = new URL("https://maps.googleapis.com/maps/api/directions/json?"
+          + "mode=driving"
+          + "&alternatives=true" + originDestination + "&key=" + key);
+    URLConnection connection = directionsUrl.openConnection();
+    connection.setRequestProperty("Accept-Charset", "UTF-8");
+    return connection;
+  }
+
+  /** 
+   * Sends request to the Routes Preferred API and returns the connection.
+   * @param origin Origin of the requested routes in "lat,lng" format.
+   * @param destination Destination of the requested routes in "lat,lng" format.
+   * @param alpha Whether to send the request to the alpha version of the API.
+   * @param key The API key.
+   */
+  public URLConnection sendPostRequestToRoutesApi(String origin, String destination, boolean alpha,
+      String apiKey) throws IOException, MalformedURLException {
+    URL routesUrl;
+    if (alpha) {
+      routesUrl = new URL("https://routespreferred.googleapis.com/v1alpha:computeCustomRoutes");
+    } else {
+      routesUrl = new URL("https://routespreferred.googleapis.com/v1:computeRoutes");
+    }
+
+    HttpURLConnection connection = (HttpURLConnection) routesUrl.openConnection();
+    connection.setRequestMethod("POST");
+    connection.setRequestProperty("Content-Type", "application/json; utf-8");
+    connection.setRequestProperty("Accept", "application/json");
+    connection.setDoOutput(true);
+
+    String requestParamsJson = "{"
+        + "\"origin\": {\"location: \"" + origin + "\"}," 
```

Comment:
I don't think this is fully right. I followed through the path through these links and it looks like you are missing the "latLng:" level, and then the format of latLng objects are {"latitude": num, "longitude":num}, but you have the simple format of "num, num". I would triple check this. (I know it's hard to check this since we can't actually run it but as best you can triple check this). 

https://developers.google.com/maps/documentation/routes_preferred/reference/rest/v1alpha/TopLevel/computeCustomRoutes#request-body
https://developers.google.com/maps/documentation/routes_preferred/reference/rest/Shared.Types/ComputeRoutesRequest#Waypoint
https://developers.google.com/maps/documentation/routes_preferred/reference/rest/Shared.Types/Location
https://developers.google.com/maps/documentation/routes_preferred/reference/rest/Shared.Types/LatLng

Reply:
yeah I fixed it and added a TODO to check with the RP team